core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala

```diff
@@ -751,6 +925,88 @@ trait CDCReaderImpl extends DeltaLogging {
     ChangeDataFeedTableFeature.metadataRequiresFeatureToBeEnabled(metadata, spark)
   }
 
+  /**
+   * Given `add` and `remove` actions of the same file, manipulate DVs to get rows that are deleted
+   * and re-added from `add` to `remove`.
+   *
+   * @return One or more [[AddFile]] and [[RemoveFile]], corresponding to CDC change_type "insert"
+   *         and "delete". Rows masked by inline DVs are changed rows.
+   */
+  private def generateFileActionsWithInlineDv(
+      add: AddFile,
+      remove: RemoveFile,
+      dvStore: DeletionVectorStore,
+      deltaLog: DeltaLog): Seq[FileAction] = {
+
+    val removeDvOpt = Option(remove.deletionVector)
+    val addDvOpt = Option(add.deletionVector)
+
+    val newActions = ListBuffer[FileAction]()
+
+    // Four cases:
+    // 1) Remove without DV, add without DV:
+    //    Not possible. This case has been handled before.
+    // 2) Remove without DV, add with DV1:
+    //    Rows masked by DV1 are deleted.
+    // 3) Remove with DV1, add without DV:
+    //    Rows masked by DV1 are added. May happen when restoring a table.
+    // 4) Remove with DV1, add with DV2:
+    //   a) Rows masked by DV2 but not DV1 are deleted.
+    //   b) Rows masked by DV1 but not DV2 are re-added. May happen when restoring a table.
+    (removeDvOpt, addDvOpt) match {
+      case (None, None) =>
+        throw new Exception("Unreadable code path reached.")
```

Comment:
nit: add few more details to the error message: ex. same file is added and removed within the same version

Reply:
Done!